### PR TITLE
[CI] Add a sanity check job  

### LINKED ISSUE #1524: [CI] Move clang-format, clang-tidy tests before tests
It would be beneficial to have the "sanity" CI checks (clang-format and clang-tidy) run as the first checks in the CI.  Currently they run after tests.  Running them before would help avoid some of the tedious churn where you submit a patch, CI chugs away for 15-ish minutes, then clang-format fails, and you have to format and re-submit.

### PR DESCRIPTION
**Pull request by uenoku**
This PR moves sanity checks(clang-format, clang-tidy, python-format) to the beginning of CI as a new job.
Will close #1524

### CODE REVIEW

File: .github/workflows/buildAndTest.yml [hunk 1]
```diff
@@ -78,7 +210,7 @@ jobs:
             -DLLVM_TARGETS_TO_BUILD=host \
             -DLLVM_ENABLE_ASSERTIONS=ON \
             -DLLVM_EXTERNAL_PROJECTS=circt \
-            -DLLVM_EXTERNAL_CIRCT_SOURCE_DIR=$PWD/.. 
+            -DLLVM_EXTERNAL_CIRCT_SOURCE_DIR=$PWD/..
```
Comment (seldridge) [inline .github/workflows/buildAndTest.yml:213]: Nit: feel free to land whitespace changes separately.

File: .github/workflows/buildAndTest.yml [hunk 2]
```diff
@@ -137,7 +269,7 @@ jobs:
 
       # Our PR builds are trying to test the two corners of the build matrix:
       # clang + release + noassert + shared
-      # gcc   + debug   + assert   + static 
+      # gcc   + debug   + assert   + static
```
Comment (seldridge) [inline .github/workflows/buildAndTest.yml:272]: Nit: feel free to land this separately.

**Review by seldridge — APPROVED** [review]
This looks great! This is going to make the PR-loop a _whole_ lot better. Thanks!

File: .github/workflows/buildAndTest.yml [hunk 3]
```diff
@@ -9,10 +9,142 @@ on:
...
+          echo "Clang-format found no formatting problems"
+          exit 0
+
+      # Run clang-tidy against only the changes. The 'clang-tidy-diff' script
+      # does this if supplied with the diff.
+      - name: clang-tidy
+        if: ${{ always() }}
+        run: |
+          git diff -U0 $DIFF_COMMIT | \
+            clang-tidy-diff -path build_assert -p1 -fix
```
Comment (teqdruid) [inline .github/workflows/buildAndTest.yml:89]: clang-tidy requires the cmake build database in `build_assert`. I assume you'll have to run a cmake configure first. (This was part of the reason I put the checks after the build to begin with.)

Comment (uenoku) [reply .github/workflows/buildAndTest.yml:89]: Thank you for pointing out. I'll fix it.

Comment (uenoku) [reply .github/workflows/buildAndTest.yml:89]: It looks currently `build_assert` doesn't exist now (for example, in https://github.com/llvm/circt/runs/4070974642?check_suite_focus=true, clang-tidy does nothing)

**Comment by uenoku** [comment]
As @teqdruid suggested, to run clang-tidy properly, it is necessary to build LLVM so clang-tidy should be putted on the last.
It will work but it creates a lot of duplication...

**Comment by uenoku** [comment]
I would like to merge this if there is a problem.

**Review by darthscsi — APPROVED** [review]
LGTM.

File: .github/workflows/buildAndTest.yml [hunk 4]
```diff
@@ -213,31 +321,13 @@ jobs:
...
-          exit 0
-
       # Run clang-tidy against only the changes. The 'clang-tidy-diff' script
       # does this if supplied with the diff.
       - name: clang-tidy
         if: ${{ always() }}
         run: |
           git diff -U0 $DIFF_COMMIT | \
-            clang-tidy-diff -path build_assert -p1 -fix
+            clang-tidy-diff -path build_clang -p1 -fix
```
Comment (teqdruid) [inline .github/workflows/buildAndTest.yml:330]: I neglected to fail this step if clang-tidy fails... You might make that change as well.

**Review by teqdruid — APPROVED** [review]
LGTM! Thanks

**Pull request merged**